### ISSUE TITLE
Inconsistent printing of fst

### ISSUE THREAD

**Issue by renkun-ken** [fstpackage/fst#131]
The `print` method of `fst` file shows the `data.frame` in an inconsistent manner after `---`:

```
> library(fst)
fst package v0.8.4
(OpenMP detected, using 8 logical cores)
> write_fst(iris, "~/data/iris.fst")
> iris_fst <- fst("~/data/iris.fst")
> iris_fst
<fst file>
150 rows, 5 columns (iris.fst)

 Sepal.Length Sepal.Width Petal.Length Petal.Width Species
          5.1         3.5          1.4         0.2  setosa
          4.9         3.0          1.4         0.2  setosa
          4.7         3.2          1.3         0.2  setosa
          4.6         3.1          1.5         0.2  setosa
          5.0         3.6          1.4         0.2  setosa
---                          
 6.7 3.0 5.2 2.3 virginica
 6.3 2.5 5.0 1.9 virginica
 6.5 3.0 5.2 2.0 virginica
 6.2 3.4 5.4 2.3 virginica
 5.9 3.0 5.1 1.8 virginica
```

**Comment by MarcusKlik** [comment]
Hi @renkun-ken, thanks for reporting, indeed, that's not too pretty!

Although it's not a critical issue, I guess it would be prudent that get a fix out soon. With this layout it's hard to see which title belongs to which column.

I will fix that ASAP!

**Comment by MarcusKlik** [comment]
Hi @renkun-ken, with commit fb5a841 I've added a nicer printing generic, one that also includes the stored table's column types:

``` r
library(fst)
library(bit64)
library(data.table)

# Sample data
x <- data.table(
  A = 1:51,
  B = sample(c(TRUE, FALSE, NA), 51, replace = TRUE),
  C = sample(1:100, 51) / 100,
  D = Sys.Date() + 1:51,
  E = sample(LETTERS, 51, replace = TRUE),
  F = factor(sample(LETTERS, 51, replace = TRUE)),
  G = as.integer64(101:151),
  I = ordered(sample(LETTERS, 51, replace = TRUE)),
  K = as.ITime(Sys.time() + 1:51))

# write to disk and then print
write_fst(x, "1.fst")
print(fst("1.fst"))
#> <fst file>
#> 51 rows, 9 columns (1.fst)
#> 
#>            A         B        C          D           E        F
#>    <integer> <logical> <double>     <Date> <character> <factor>
#> 1          1     FALSE     0.44 2018-01-28           K        V
#> 2          2     FALSE     0.83 2018-01-29           K        E
#> 3          3      TRUE     0.33 2018-01-30           H        P
#> 4          4        NA     0.85 2018-01-31           J        H
#> 5          5     FALSE     0.38 2018-02-01           O        P
#>          ---                                                   
#> 47        47        NA     0.11 2018-03-15           H        I
#> 48        48      TRUE     0.94 2018-03-16           Z        L
#> 49        49     FALSE     0.15 2018-03-17           O        Y
#> 50        50      TRUE     0.79 2018-03-18           Q        D
#> 51        51        NA     0.49 2018-03-19           D        F
#>              G                I        K
#>    <integer64> <ordered factor>  <ITime>
#> 1          101                J 23:57:31
#> 2          102                K 23:57:32
#> 3          103                V 23:57:33
#> 4          104                R 23:57:34
#> 5          105                C 23:57:35
#>                                         
#> 47         147                Z 23:58:17
#> 48         148                S 23:58:18
#> 49         149                O 23:58:19
#> 50         150                D 23:58:20
#> 51         151                R 23:58:21
```

The `data.table` dots are there, as are descriptors for all supported types in `fst`.  I hope this looks better, all feedback very welcome!

**Comment by MarcusKlik** [comment]
(I'm still looking into how to add the `RStudio` coloring, like in a `tibble`, that would be a nice extra feature):

![image](https://user-images.githubusercontent.com/24658401/35465906-c3cc86c0-02ff-11e8-81af-b7cff4bb2fbe.png)

**Comment by renkun-ken** [comment]
Nice work! In my knowledge, the coloring is done by `crayon` package, which directly works in a typical terminal. RStudio just updates its web terminal component to support displaying colored texts. 

A simple example in RStudio is

```r
cat(crayon::red("hello\n"))
```

<img width="210" alt="image" src="https://user-images.githubusercontent.com/4662568/35467163-266996ba-0346-11e8-8724-05e7be078b0f.png">

`crayon` package makes it easier to add context information to the text to print. Without it, the coloring also works:

```r
cat("\033[31mhello\n\033[39m")
```

**Comment by MarcusKlik** [comment]
Thanks for the pointers @renkun-ken! I used the special codes from the `crayon` package to give the row names and the '--' separators a light grey color. The column types are displayed in light grey as well but are in italics. This is how the table from the example above looks in RStudio now:

![image](https://user-images.githubusercontent.com/24658401/35488243-a7c1bd74-0485-11e8-8da8-042563dfe272.png)

The colors are identical to those used in the `tibble` printing output and I used the separator from `data.table` (but with a color).

**Comment by renkun-ken** [comment]
Looks nice!

**Comment by MarcusKlik** [comment]
Thanks! I will close the issue for now, the new formatting will be available in the next `fst` release (`v0.8.6`). Note that the current implementation can probable be made faster by using overrides to the `format` generic. Currently the existing `data.frame` implementation for printing is leveraged, using `capture.output` after which the headers and coloring are applied to the captured output.
